Valuetype refactor

3starblaze opened the pull request.

Commit:
Turn ValueTypes into classes

Commit:
Rename toJson to toJSON

Commit:
Fix breaking changes from modifying ValueType

Commit:
Resolve errors and tests

Commit:
Create Value class and standard values

Commit:
Add "nothing" to Value

Commit:
Make Value wrap any* value not JsonValue

* Not really any, the ones that can be converted to JSON.

Also includes improved toJSON().

Commit:
Adjust value checking process

Commit:
Fix standard value definitions

Commit:
Add value set/get and setNothing()

Commit:
Remove "nothing" from Value

Commit:
"nothing" and "isSet" changes

From now on no value object implies nothing state which shows the mutual
exclusion between having a value and having "nothing".

hasValue now is renamed to "isSetVariable" to indicate the similarity
with "isSet" and to reduce ambiguity of "hasValue" (which is not an
exactly correct term)

Missing visibility modifiers have been added for explicitness.

Commit:
Add setNothing()

Commit:
Partially update inputSocket

In a similar way Socket was changed recently

Commit:
Fix Outputsocket like Socket

Commit:
Fix remaining socket issues

This includes fixing the return type of getters.

Commit:
Fix some issues caused by missing generic types

unknown is passed as generic because a socketCollection can consist of
sockets with different types

Commit:
Add some detail to isSetVariable documentation

This includes the importand distinction of value = null (aka nothing)
and isSetVariable = false

Commit:
Implement valueDefinitions

Commit:
Create standard value definitions

Commit:
Move ValueConstructor to value.ts

Commit:
Fix some inconsistencies

Commit:
Rename unset(reset) to clear and init to reset

"unset" was "reset" before but it should have been called "clear". This
includes not only changes in sockets but also node and network which
followed a similar naming convention.

Commit:
Swap unknown assignments to Value<unknown>

Commit:
Make sockets demand Value constructors

This new change makes it possible to declare the type without passing a
value. Also users don't have to wrap a value with Value<T> which makes
the sockets easy to use. This is made even easier in Node where addInput
and addOutput ask for name which then is used to retrieve a
corresponding Value constructor (or signal an error).

Some changes made in nodeTypeDefinitions to reflect the new API. It's
not transformed fully because there is no standard Value type yet.

Commit:
Add default Value type

Commit:
Fix tests and add tiny features

Features include setDefaultNothing for inputSocket and isWaiting for
outputSocket.

Tests also include examples. The quadraticForumalaNetworkTest was
suspended because it's bloated.

Commit:
Fix node reset() recursion

Commit:
Delegate default value type business to socket

This is done so that socket can be used without an engine.
commit WIP

Commit:
Make node ignore engine if no type name provided

Before this commit Node always checked for engine but for default
type (when name is omitted) the engine is not necessary.

A helper has been created to avoid duplicating code in addInput and addOutput.

Commit:
Fix linter issues

Commit:
Remove ValueType

Passiday commented on src/Dynode/controller/network.ts:
```diff
@@ -24,13 +24,13 @@ export default class NetworkController {
           };
           if (socket.isDefaultSet() && !socket.isDefaultNothing()) {
-            result.value = socket.getJsonDefaultValue();
+            result.value = socket.getValue().toJSON();
```
Here the default value has to be passed to the UI component, instead of value.

Passiday commented on src/Dynode/model/core/node.ts:
```diff
@@ -64,7 +66,24 @@ class Node extends VEventTarget {
+   *
+   * @param name ValueConstructor name that is stored in engine
+   */
+  private valueTypeExtractor(name?: string): ValueConstructor<unknown> | undefined {
```
getValueType()

Passiday commented on src/Dynode/model/core/node.ts:
```diff
@@ -115,7 +127,7 @@ class Node extends VEventTarget {
-  getInputValue(name: string): unknown {
+  getInputValue(name: string): Value<unknown> {
     const input = this.getInput(name);
     return input.getValue();
```
What if we pass `getValue().value` here? In the model context, this is pretty much the only thing we care when we ask for an input value. In the rare case when the check/JSON methods are needed, we can always use `getInput(name)`.

3starblaze replied:
Since Value class is going to be renamed, socket's getValue should also reflect the same nomenclature. I would say that getInputValue should in fact return raw value, though input.getValue() should do the same (otherwise it should be named differently)

Passiday commented on src/Dynode/model/core/socket/inputSocket.ts:
```diff
@@ -1,97 +1,92 @@
    */
-  isValid(): boolean {
-    return this.linkedSocket ? true : this.hasDefault;
+  public isValid(): boolean {
```
If we agree that unlinked input sockets with no default set should be treated as nothing, this method is no longer needed. But it is important to make sure that this behavior is properly implemented.

3starblaze replied:
This is outside this PR's scope because it's not related to ValueType and the PR is already rather big (which would further on delay merging this to master).

Passiday commented on src/Dynode/model/core/socket/inputSocket.ts:
```diff
@@ -128,30 +123,25 @@ class InputSocket extends Socket {
       return super.isSet();
     }
-    return this.hasDefault;
+    return this.isDefaultSet();
```
If we treat unlinked input sockets as nothing, then unlinked input sockets are always set. So, just return true here.

Passiday commented on src/Dynode/model/core/socket/inputSocket.ts:
```diff
@@ -1,97 +1,92 @@
-    if (!this.hasDefault) throw Error('Input socket has no default value set');
-    return this.defaultNothing;
+  public isDefaultNothing(): boolean {
+    if (!this.isDefaultSet) throw Error('Input socket has no default value set');
```
if default is not set, treat is as nothing state. So, just return true here, instead of throwing an exception.

Passiday commented on src/Dynode/model/core/socket/outputSocket.ts:
```diff
@@ -37,75 +32,82 @@ class OutputSocket extends Socket {
    */
-  constructor(parentNode: Node, typeObject?: ValueType, storageMode?: boolean) {
-    super();
+  constructor(parentNode: Node, value?: T, ValueType?: ValueConstructor<T>, storageMode?: boolean) {
```
Here you are using "ValueType" to denote the related value type class constructor. I support this naming convention, but you have to be consistent how you name these references to value type class constructors.

3starblaze replied:
valueType and ValueConstructor would be renamed SocketValue and SocketValueType accordingly.

Passiday commented on src/Dynode/model/core/socket/socket.ts:
```diff
@@ -1,109 +1,109 @@
+  public title: string | null = null;
 
-  constructor(typeObject?: ValueType) {
+  constructor(value?: T, ValueType?: ValueConstructor<T>) {
```
I think, we don't need to pass the socket value at construction time. Normally, the value is set when the network is resolved. Also, if a value is set at construction time, nobody can subscribe to the socket's "value" event to catch the value update.

3starblaze replied:
Agreed, I'll remove value from the constructor.

Commit:
Fix bug where default value should have been taken

Commit:
Rename ValueTypeExtractor to getValueType

Commit:
Rename Value(Constructor) to SocketValue(Type)

Commit:
Remove value from socket constructor + minor fixes

This change also affects Node's addInput and addOutput that mirror
socket constructor.

Minor fixes include changing Value(Constructor) to SocketValue(Type).

Commit:
Change old .getInput().value to getInput()

Some of them are not fixed because they actually need access to that SocketValue object

Commit:
Implement getSocketValue

Commit:
Fix minor linter and naming issues

Passiday commented on src/Dynode/model/core/node.ts:
```diff
@@ -64,25 +66,35 @@ class Node extends VEventTarget {
+   *
+   * @param name ValueSocketType name that is stored in engine
+   */
+  private getValueType(name?: string): SocketValueType<unknown> | undefined {
```
getSocketValueType()

Passiday replied:
But I think it is a bit awkward that Node class needs this helper function. Since the types live in the engine, it would be more natural to do something like this.engine.getSocketValueType(), instead of introducing this helper function. One of the reasons is that node is not aware of the network it is part of, and access to the engine should be through the associated network. And fixing that would indeed be something that has to be done in a separate PR.

Passiday commented on src/Dynode/model/core/engine.ts:
```diff
@@ -22,17 +22,9 @@ class Engine {
-      if (v.name === valueType) return v;
-    }
-    throw Error(`Type ${valueType} does not exist!`);
+  public getSocketValueTypeDefinition(name: string): SocketValueType<unknown> {
```
getSocketValueType()

Passiday commented on src/Dynode/model/core/engine.ts:
```diff
@@ -1,8 +1,8 @@
 
 class Engine {
-  private valueTypeDefinitions: ValueType[] = [];
+  private socketValueTypeDefinitions: Record<string, SocketValueType<unknown>> = {};
```
socketValueTypes

Passiday commented on src/Dynode/model/core/engine.ts:
```diff
@@ -42,23 +34,20 @@ class Engine {
-      if (v.name === valueType.name) {
-        throw Error(`ValueType ${v.name} already exists!`);
-      }
+  public addSocketValueTypeDefinition(
```
addSocketValueType()

Commit:
Test various socket values and fix isJSON()

Commit:
Test and tweak NumberSocketValue

This also includes some modifications in SocketValue that calls static
methods dynamically (it uses this.constructor to find the classname).

Right now NaN and +-Infinity is accepted at number but not in normal SocketValue.

Commit:
Fix broken tests by fixing e.target.getValue()

The problem is that sockets were unexpectedly passed undefined, even
though they had actual values. After skimming through the stack trace I
noticed getValue().value which shouldn't be correct and as I discovered
e.target is of type any.

Commit:
Create a folder for socket tests

Commit:
Test default socket type assignment

Commit:
Fix linter issues

Commit:
Pass values as strings as a temporary solution

Commit:
Fix the async test
